resource/planner/mintime_resource_tree.cpp

```diff
@@ -0,0 +1,272 @@
+    return nullptr;
+}
+
+int64_t mintime_resource_tree_t::find_mintime_anchor (
+            int64_t request, mt_resource_rb_node_t **anchor_p)
+{
+    mt_resource_rb_node_t *node = m_tree.get_root ();
+    int64_t min_time = std::numeric_limits<int64_t>::max ();
+    int64_t right_min_time = std::numeric_limits<int64_t>::max ();
+    while (node) {
+        if (request <= node->remaining) {
+            // visiting node satisfies the resource requirements. This means all
+            // nodes at its subtree also satisfy the requirements. Thus,
+            // right_min_time is the best min time.
+            right_min_time = right_branch_mintime (node);
+            if (right_min_time < min_time) {
+                min_time = right_min_time;
+                *anchor_p = node;
+            }
+            // next, we should search the left subtree for potentially better
+            // then current min_time;
+            node = node->get_left ();
+        } else {
+            // visiting node does not satisfy the resource requirements. This
+            // means that nothing in its left branch will meet these requirements:
+            // time to search the right subtree.
+            node = node->get_right ();
+        }
+    }
+    return min_time;
+}
+
+
+/*******************************************************************************
+ *                                                                             *
+ *                 Private Mintime Resource RBTree Node Methods                *
+ *                                                                             *
+ *******************************************************************************/
+
+template <class mt_resource_rb_node_t, class NodeTraits>
+void mt_resource_node_traits<mt_resource_rb_node_t, NodeTraits>::fix_propagate (
+         mt_resource_rb_node_t *node)
+{
+    while ( (node = node->get_parent()) != nullptr) {
+        int64_t min = node->at;
+        mt_resource_rb_node_t *left = node->get_left ();
+        if (left) {
+            if (min > left->subtree_min)
+                min = left->subtree_min;
+        }
+        mt_resource_rb_node_t *right = node->get_right ();
+        if (right) {
+            if (min > right->subtree_min)
+                min = right->subtree_min;
+        }
+        if (node->subtree_min == min)
+            break;
+        node->subtree_min = min;
+    }
+}
+
+template <class mt_resource_rb_node_t, class NodeTraits>
+void mt_resource_node_traits<mt_resource_rb_node_t, NodeTraits>::fix (
+         mt_resource_rb_node_t *node)
+{
+    int64_t min = node->at;
+    mt_resource_rb_node_t *left = node->get_left ();
+    if (left) {
+        if (min > left->subtree_min)
+            min = left->subtree_min;
+    }
+    mt_resource_rb_node_t *right = node->get_right ();
+    if (right) {
+        if (min > right->subtree_min)
+            min = right->subtree_min;
+    }
+    node->subtree_min = min;
+    fix_propagate (node);
```

Comment:
Nit, but I don't think `fix_propagate` is strictly necessary. I think you could pull this off in one function with less code duplication (totally optionatl):

```suggestion
while(node != nullptr) {
    int64_t min = node->at;
    mt_resource_rb_node_t *left = node->get_left ();
    if (left) {
        if (min > left->subtree_min)
            min = left->subtree_min;
    }
    mt_resource_rb_node_t *right = node->get_right ();
    if (right) {
        if (min > right->subtree_min)
            min = right->subtree_min;
    }
    node->subtree_min = min;
    node = node->get_parent();
}
```

Reply:
> I think you could pull this off in one function with less code duplication (totally optionatl):

BTW, code duplication is a part of initial performance optimization. This is the hottest spot of mintime resource search tree. Although modern C++ compilers do a good job of inlining, I found non negligible overheads with an old abstraction like `mintime_resource_subtree_min` in a separate function.

Reply:
I know where you are coming from. But the loop body of the `fix_propagate` and the node `fix` code is slightly different. The `fix_propagate` has one conditional statement where terminates the loop when the visiting node's existing `subtree_min` is the same as the new min. This way, you don't have to propagate the fixup to your ancestor nodes unless you have to.

How I see this is, there is one special case -- the initial node where the fixup is applied and then general case where the fixup + a conditional is performed. So you can consider this a loop unrolling: a bit better performance and a bit better readability,

Hopefully this makes sense.

Reply:
Actually, after looking at your downstream comments, I have a second thought. Perhaps implement the whole thing in a single loop makes more sense for readability. I will mark it resolved here and describe my thinking in one of your other comments below.